baseapp/saml/serviceprovider.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2019 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package saml
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/RobotsAndPencils/go-saml"
+	"github.com/pkg/errors"
+)
+
+const (
+	HTTPRedirectBinding = "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
+)
+
+type OnErrorCallback func(*http.Request, error)
+
+type OnLoginCallback func(http.ResponseWriter, *http.Request, *saml.Response)
+
+type Settings struct {
+	AssertionConsumerServicePath string
+	SignRequests                 bool
+	SigningCertificatePath       string
+	SigningKeyPath               string
+	IDPMetadataURL               string
+	OnError                      OnErrorCallback
+	OnLogin                      OnLoginCallback
+}
+
+// ServiceProvider is capable of handling a SAML login. It provides
+// an http.Handler (via ACSHandler) which can process the http POST from the SAML IDP. It accepts callbacks for both error and
+// success conditions so that clients can take action after the auth flow is complete. It also provides a handler
+// for serving the service provider metadata XML.
+type ServiceProvider struct {
+	settings *saml.ServiceProviderSettings
+	acsPath  string
+	onError  OnErrorCallback
+	onLogin  OnLoginCallback
+}
+
+// NewServiceProviderFromMetadata returns a ServiceProvider. The configuration of the ServiceProvider
+// is a result of combinging settings provided to this method and values parsed from the IDP's metadata.
+func NewServiceProviderFromMetadata(settings Settings) (*ServiceProvider, error) {
+	s, err := getSettingsFromMetadata(settings)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not determine settings from IDP metadata")
+	}
+	sp := &ServiceProvider{
+		settings: s,
+		onError:  settings.OnError,
+		onLogin:  settings.OnLogin,
+		acsPath:  settings.AssertionConsumerServicePath,
+	}
+
+	if sp.onError == nil {
+		sp.onError = DefaultErrorCallback
+	}
+
+	if sp.onLogin == nil {
+		sp.onLogin = DefaultLoginCallback
+	}
+
+	return sp, nil
+}
+
+func DefaultErrorCallback(r *http.Request, err error) {
+	fmt.Println(err.Error())
+}
+
+func DefaultLoginCallback(w http.ResponseWriter, r *http.Request, resp *saml.Response) {
+	w.WriteHeader(http.StatusOK)
+}
+
+func getSettingsFromMetadata(settings Settings) (*saml.ServiceProviderSettings, error) {
+	resp, err := http.Get(settings.IDPMetadataURL)
```

Comment:
`getSettingsFromMetadata` could also take a `[]byte` so that users could either provide file contents or the url.